coverage-tool/coverage-tool.py

```diff
@@ -0,0 +1,300 @@
+    doc_path = os.path.join(SCRIPT_DIR, "discovery_doc.json")
+    if not os.path.exists(doc_path):
+        with open(doc_path, "w") as doc:
+            with urlopen(DISCOVERY_DOC_URL) as response:
+                doc.write(response.read().decode("utf-8"))
+    return json.load(open(doc_path))["schemas"]
+
+
+def get_mock_responses_list():
+    """Return a list of mock response files to evaluate according to args."""
+    mock_responses = sorted(os.listdir(MOCK_RESPONSES_PATH))
+    if args.exclude:
+        mock_responses = [f for f in mock_responses if f not in args.exclude]
+    if args.scan_files:
+        content = ""
+        for file_path in args.scan_files:
+            with open(file_path) as f:
+                content += f.read()
+        mock_responses = [f for f in mock_responses if f.split(".")[0] in content]
+    return mock_responses
+
+
+def load_mock_response(file_path):
+    """Return a mock response file's content as a list of dictionaries."""
+    with open(file_path) as f:
+        lines = f.readlines()
+    ans = []
+    for n, line in enumerate(lines):
+        if line.strip() == "":
+            continue
+        elif line.startswith("data:"):
+            # For streaming responses
+            ans.append(json.loads(line[5:]))
+        else:
+            # For unary responses and last part of some streaming responses
+            ans.append(json.loads("".join(lines[n:])))
+            break
+    return ans
+
+
+def is_response_type(file_part, schema_props):
+    """Check if the given response file part matches the given schema properties."""
+    for field in file_part:
+        if field not in schema_props:
+            return False
+        if "$ref" in schema_props[field]:
+            # Recursively check nested properties
+            if not is_response_type(
+                file_part[field], schemas[schema_props[field]["$ref"]]["properties"]
+            ):
+                return False
+    return True
+
+
+def get_grouped_mock_responses():
+    """Return a dictionary of mock responses grouped by response type."""
+    responses_content = {
+        file_name: load_mock_response(os.path.join(MOCK_RESPONSES_PATH, file_name))
+        for file_name in get_mock_responses_list()
+    }
+    grouped_mock_responses = {t: {} for t in RESPONSE_TYPES}
+    # Check each file's first part against each response type
+    for file_name, content in responses_content.items():
+        for response_type in RESPONSE_TYPES:
+            response_type_props = schemas[SCHEMA_PREFIX + response_type]["properties"]
+            if is_response_type(content[0], response_type_props):
+                grouped_mock_responses[response_type][file_name] = content
+                break
+        else:
+            # Failure responses are not expected to match any response type
+            if "failure" not in file_name:
+                raise Exception(f"File {file_name} does not match any response type.")
```

Comment:
Can you use a more specific exception

Reply:
We're printing a warning instead because files like [unary-failure-invalid-response.json](https://github.com/FirebaseExtended/vertexai-sdk-test-data/blob/main/mock-responses/unary-failure-invalid-response.json) and [streaming-failure-invalid-json.txt](https://github.com/FirebaseExtended/vertexai-sdk-test-data/blob/main/mock-responses/streaming-failure-invalid-json.txt) don't match any response type by design.